refactor(pool/replica): add factories for pool&replica

tiagolobocastro opened the pull request:
This allows us to replace the existing reuse of the gRPC trait and allows for a easier addition of another backend type.
New backends should impl at least PoolOps, PoolProps, ReplicaOps and LogicalVolume. todo: snapshot is a separate service, which would still require refactoring.

tiagolobocastro committed:
refactor(pool/replica): add factories for pool&replica

This allows us to replace the existing reuse of the gRPC trait and allows for
a easier addition of another backend type.
New backends should impl at least PoolOps, PoolProps, ReplicaOps and LogicalVolume.
todo: snapshot is a separate service, which would still require refactoring.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

dsharma-dc commented on io-engine/src/core/logical_volume.rs:
```diff
@@ -1,34 +1,53 @@
-    /// Returns the name of the Logical Volume
+use crate::{core::Protocol, pool_backend::PoolBackend};
+
+///  LogicalVolume Trait Provide all the Generic Interface for Volume.
```
```suggestion
///  LogicalVolume Trait Provide all the Generic Interface for a Logical Volume on any backend type.
```

dsharma-dc commented on io-engine/src/core/logical_volume.rs:
```diff
@@ -1,34 +1,53 @@
+    fn snapshot_uuid(&self) -> Option<String>;
+
+    /// Return the share protocol of this Logical Volume.
+    fn share(&self) -> Protocol;
```
Let's call this `share_proto` or `shared_over` , otherwise the name conflicts with the `share` verb function.

dsharma-dc commented on io-engine/src/grpc/v1/lvm/pool.rs:
```diff
@@ -211,17 +31,8 @@ impl From<LvmError> for Status {
-        }
+    fn from(vg: VolumeGroup) -> Self {
+        let lvs = &vg as &dyn PoolProps;
+        lvs.into()
```
```suggestion
        let lvm_pool = &vg as &dyn PoolProps;
        lvm_pool.into()
```

dsharma-dc commented on io-engine/src/grpc/v1/pool.rs:
```diff
@@ -327,31 +352,152 @@ impl PoolService {
+                let lock_guard =
+                    acquire_subsystem_lock(pool_subsystem, Some(pool.name()))
+                        .await?;
+                Ok(Some(PoolGrpc::new(pool, lock_guard)))
```
What is the lifetime of PoolGrpc object ? The guard is tied to that.

tiagolobocastro replied:
Can think of it as : 'static, which is, until `PoolGrpc` is dropped, example:
```rust
crate::spdk_submit!(async move '1: {
      let pool: PoolGrpc  = ...;
      pool.create_replica(args).await    // <- when create_replica completes and scope '1 is complete then PoolGrpc is dropped
  })
```

dsharma-dc reviewed (commented):
halfway...

dsharma-dc commented on io-engine/src/core/logical_volume.rs:
```diff
@@ -1,34 +1,54 @@
+    /// Returns a boolean indication if the Logical Volume is a clone.
+    fn is_clone(&self) -> bool;
+
+    /// Returns its parent snapshot uuid if the Logical Volume is a snapshot.
```
```suggestion
    /// Returns its parent snapshot uuid if the Logical Volume is a clone.
```

dsharma-dc commented on io-engine/src/grpc/v1/pool.rs:
```diff
@@ -447,59 +575,46 @@ impl PoolRpc for PoolService {
+
+                    // todo: what is the intent here when None, to only return
+                    // pools  of Lvs?
+                    // todo: Also, what todo when we hit an error listing any of
```
My suggestion is, for listing, let's not be concerned about pool type specific behaviour. Consider to list pools of all types, and if there is any error listing something then let it be reported as such.

tiagolobocastro replied:
how do we report an error, output to stdout is enough?

dsharma-dc replied:
> how do we report an error, output to stdout is enough?

Yeah, perhaps add a log or an event as well? I wouldn't think it's possible  to send a partial list and and error code too indicating that it's incomplete.

dsharma-dc commented on io-engine/src/grpc/v1/replica.rs:
```diff
@@ -277,17 +337,34 @@ where
+            || pool.uuid() != replica.pool_uuid()
+        {
+            let msg = format!(
+                "Specified pool: {pool:?} does match the target replica: {replica:?}!"
```
```suggestion
                "Specified pool: {pool:?} does not match the target replica's pool: {replica:?}!"
```

dsharma-dc commented on io-engine/src/lvm/lv_replica.rs:
```diff
@@ -869,6 +866,34 @@ impl crate::core::LogicalVolume for LogicalVolume {
         }
     }
+
+    fn is_snapshot(&self) -> bool {
```
Do we somehow disallow snapshots on lvm LVs today? (I'm guessing yes because we haven't implemented `--snap` support yet?). If yes then returning this blanket `false` looks ok, otherwise we need some logic here.

tiagolobocastro replied:
Correct, not implemented :)
Also the snapshot code is atm all lvs specific, would have to modify traits for snapshots too :(
Can be taken as separate effort

dsharma-dc commented on io-engine/src/replica_backend.rs:
```diff
@@ -18,39 +17,121 @@ pub trait ReplicaOps: Share + LogicalVolume {
+}
+
+#[macro_export]
+macro_rules! repl_factory {
```
I am confused here! Why are we implementing ReplicaFactory via macro here, and we have implemented it already as:
> impl ReplicaFactory<LogicalVolume, Error> for ReplLvmFactory
and
> impl ReplicaFactory<Lvol, LvsError> for ReplLvsFactory

tiagolobocastro replied:
This was to automatically generate a factory with the grpc types, example: `PoolFactory<Pool, tonic::Status>`
I have now made the trait more generic, and grpc conversion implemented without a trait so these macros will all be removed.

dsharma-dc reviewed (approved):
This refactoring is worth documenting(the class/interface details) on confluence for benefit and clarity of everyone.

Abhinandan-Purkait commented on io-engine/src/core/logical_volume.rs:
```diff
@@ -1,34 +1,54 @@
+    fn snapshot_uuid(&self) -> Option<String>;
+
+    /// Return the share protocol of this Logical Volume.
+    fn share_proto(&self) -> Protocol;
```
`share_proto` --> `share_protocol` proto is a different entity even though it might not be related to this.

Abhinandan-Purkait commented on io-engine/src/lvm/lv_replica.rs:
```diff
@@ -16,6 +16,7 @@ use crate::{
     },
 };
 
+use crate::pool_backend::PoolBackend;
```
Move with line 2?

tiagolobocastro commented:
bors merge

tiagolobocastro committed:
ci(python/bdd): improve xunit reports

Rather than keeping them flat in the workspace, place them on a specific
folder (test/python/reports).

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented:
bors try

tiagolobocastro commented:
bors try

tiagolobocastro commented:
bors try

tiagolobocastro commented:
bors merge
